pkg/util/hack/map_abi.go

```diff
@@ -0,0 +1,435 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hack
+
+import (
+	"runtime"
+	"strings"
+	"unsafe"
+)
+
+// Maximum size of a table before it is split at the directory level.
+const maxTableCapacity = 1024
+
+// Number of bits in the group.slot count.
+const swissMapGroupSlotsBits = 3
+
+// Number of slots in a group.
+const swissMapGroupSlots = 1 << swissMapGroupSlotsBits // 8
+
+// src/internal/runtime/maps/table.go:table
+type swissMapTable struct {
+	// The number of filled slots (i.e. the number of elements in the table).
+	used uint16
+
+	// The total number of slots (always 2^N). Equal to
+	// `(groups.lengthMask+1)*abi.SwissMapGroupSlots`.
+	capacity uint16
+
+	// The number of slots we can still fill without needing to rehash.
+	//
+	// We rehash when used + tombstones > loadFactor*capacity, including
+	// tombstones so the table doesn't overfill with tombstones. This field
+	// counts down remaining empty slots before the next rehash.
+	growthLeft uint16
+
+	// The number of bits used by directory lookups above this table. Note
+	// that this may be less then globalDepth, if the directory has grown
+	// but this table has not yet been split.
+	localDepth uint8
+
+	// Index of this table in the Map directory. This is the index of the
+	// _first_ location in the directory. The table may occur in multiple
+	// sequential indicies.
+	//
+	// index is -1 if the table is stale (no longer installed in the
+	// directory).
+	index int
+
+	// groups is an array of slot groups. Each group holds abi.SwissMapGroupSlots
+	// key/elem slots and their control bytes. A table has a fixed size
+	// groups array. The table is replaced (in rehash) when more space is
+	// required.
+	//
+	// TODO(prattmic): keys and elements are interleaved to maximize
+	// locality, but it comes at the expense of wasted space for some types
+	// (consider uint8 key, uint64 element). Consider placing all keys
+	// together in these cases to save space.
+	groups groupsReference
+}
+
+// groupsReference is a wrapper type describing an array of groups stored at
+// data.
+type groupsReference struct {
+	// data points to an array of groups. See groupReference above for the
+	// definition of group.
+	data unsafe.Pointer // data *[length]typ.Group
+
+	// lengthMask is the number of groups in data minus one (note that
+	// length must be a power of two). This allows computing i%length
+	// quickly using bitwise AND.
+	lengthMask uint64
+}
+
+// src/internal/runtime/maps/map.go:Map
+// Note: changes here must be reflected in cmd/compile/internal/reflectdata/map_swiss.go:SwissMapType.
+type swissMap struct {
+	// The number of filled slots (i.e. the number of elements in all
+	// tables). Excludes deleted slots.
+	// Must be first (known by the compiler, for len() builtin).
+	Used uint64
```

Comment:
I can't find the code that updates `Used` field in tidb. How it is updated?

Reply:
Actually, this filed is the [used](https://github.com/golang/go/blob/go1.25.2/src/internal/runtime/maps/map.go#L199) field of the golang internal map (read only for TiDB). It will be modified while insert/del elements towards map.

The `func ToSwissMap` is used to decode a internal map by same abi.

Reply:
> Actually, this filed is the [used](https://github.com/golang/go/blob/go1.25.2/src/internal/runtime/maps/map.go#L199) field of the golang internal map (read only for TiDB). It will be modified while insert/del elements towards map.
> 
> The `func ToSwissMap` is used to decode a internal map by the same abi.

Once the structure of the golang internal map is changed, can existing tests ensure that they will fail?